### DISCUSSION TITLE
[Request] Minimum Login Time for Claim Auto-Expire Reset

### DISCUSSION THREAD

**Discussion by ghost** [GriefPrevention/GriefPrevention#58] [Ideas]
Hello! I have a feature request for you all.

I would like to be able to configure a minimum amount of time a player must be logged in for their claims to be considered active.
My issue is players will just login for ten seconds and log back out just so they can prevent their claims from expiring. While this preserves their builds, the idea is that inactive players claims should expire so others can loot, explore or claim them for their own.
I'm thinking it would work like adding up total time logged in within a specific timeframe (say in one day or a week or soemthing) and letting the server admin set a minimum amount of time in minutes or 0 to disable.
How you wish to implement this of course, if at all, is up to you! Thank you for reading.

**Comment by plasticblues** [comment]
I really like this idea, I've played on a few servers with this plugin and this was happening there as well, in fact I even exploited this before, login for 2 seconds just to preserve my claim then logout and not play again until I knew my claim was going to expire. I could hold on to my claims for years or until I felt like playing on the server again.

**Comment by RoboMWM** [comment]
GP used to use its own system for logging the last time the player was on the server - however, we now use Bukkit's API to get that info now (thanks to @Jikoo - and I kid of course).

Thing is, this may stop some of these players - but what if players, instead of logging in for 10 seconds, just sit in the server until they're AFK-kicked?

This whole "AFK" thing has been on my mind (in general), and although there's way to combat basic types, there's clients that can have the player doing normal player things while the person is actually AFK, so for the most part I see it as a largely-fruitless endeavor.

Also, is there something about these claims that you'd like to see such being abandoned? (This is more just me trying to understand why you guys really want claims of "inactive" players to be abandoned even if they might've built cool things, idk.)

General idea beside, there could be ways you could combat this - if you notice some players doing this (which is obviously why you're reporting this here, otherwise I doubt you'd notice), why not just ban them/abandon their claims?

**Comment by ghost** [comment]
@RoboMWM I use a combination of plugins to combat AFK players. Essentials, Autorank and afkTerminator. 
Firstly, Essentials will detect if a player is AFK and automatically flag them as such if they are inactive for 15 minutes or more. Autorank hooks into essentials and stops counting their playtime when this happens. Perhaps this same hook could be added for GP?
If a player avoids this AFK detection, afkTerminator steps in. It detects if a player is using a variety of techniques to avoid detection. If the player is suspected of avoiding AFK detection, they are kicked and automatically get playtime removed from their Autorank playtime as well as a message sent via Essentials mail about what happened.
afkTerminator can also detect people who auto re-log by forcing players to wait a certain number of seconds before re-logging (an interval that increase with the number of too-quick attempts), eventually resulting in a temporary 30 minute ban. Read more on afkTerminator here: (they have an API) https://dev.bukkit.org/projects/afkterminator 

As for wanting the claims to be abandoned, I use claims through an economy system. Players buy claimblocks to protect land. As long as they're active their builds are safe. If they are inactive for a certain number of days the claims get removed and other active or new players can do what they want with the abandoned creations. Loot, salvage, claim for safekeeping, or enhance; it's up to them. Basically as players come and go the server fills with ruins of players past and it creates an interesting environment that the players control entirely.

I could combat this by checking to see who does this and banning them, but honestly I'm just plain lazy! I don't really monitor my server too closely, it's more of a thing I check up on every now again or check if someone tells me it's on fire. I rarely intervene with the ongoings of the server. But a way to prevent players from doing this in the first place would make it way easier! No idea on the difficulty on actually implementing it though...

**Comment by Jikoo** [comment]
My issue with how GP was tracking logins before was that sometimes player files wouldn't properly update and I was having players' claims expire early despite them having logged in since. Using Bukkit's API for last login was a much safer option when GP would occasionally not save the date correctly. I have no idea why, unfortunately - my expiration is at 5 months, and the player who got me noticing it being an issue in the first place had been offline for 2. By now I don't have logs of it, no idea if there was a crash that day or something. Unfortunately, it wasn't an isolated issue. I found another couple dozen being expired early in the following weeks once I was aware that it might need double checking.

Definitely a good feature, but I think it needs to be approached with caution. In my case, being a tiny community, I'd prefer it to use the safe route of checking the Bukkit API when not set to require a minimum session duration. Larger servers I'd imagine are much less concerned about little edge cases slipping through the cracks, anyone who'd fall victim to expiration is hardly a regular anyway.

**Comment by RoboMWM** [comment]
@Artige I'm fully aware of the AFK plugins out there. I was aware of AFK terminator years ago. All of these solutions primarily assume the client is vanilla.

I'm assuming the API allows for claims to be deleted - and as such, if someone wants more robust "claim expiration," I believe an addon can fill for this.

**Comment by BillyGalbreath** [comment]
I'll make an addon for this because its something I would use myself, too.

**Comment by ghost** [comment]
@BillyGalbreath Awesome!
